apps/rush-lib/src/logic/PackageJsonUpdater.ts

```diff
@@ -111,66 +116,63 @@ export class PackageJsonUpdater {
    * Adds a dependency to a particular project. The core business logic for "rush add".
    */
   public async doRushAdd(options: IPackageJsonUpdaterRushAddOptions): Promise<void> {
-    const {
-      projects,
-      packageNames,
-      initialVersions,
-      devDependency,
-      updateOtherPackages,
-      skipUpdate,
-      debugInstall,
-      rangeStyles,
+    const { projects, packagesToAdd, devDependency, updateOtherPackages, skipUpdate, debugInstall, variant } =
+      options;
+
+    const allDependencyVersions: Map<string, Set<string>> = this._rushConfiguration._getAllDependencyVersions(
       variant
-    } = options;
-
-    const implicitlyPinned: Map<string, string> =
-      this._rushConfiguration.getImplicitlyPreferredVersions(variant);
-    const purgeManager: PurgeManager = new PurgeManager(this._rushConfiguration, this._rushGlobalFolder);
-    const installManagerOptions: IInstallManagerOptions = {
-      debug: debugInstall,
-      allowShrinkwrapUpdates: true,
-      bypassPolicy: false,
-      noLink: false,
-      fullUpgrade: false,
-      recheckShrinkwrap: false,
-      networkConcurrency: undefined,
-      collectLogFile: false,
-      variant: variant,
-      maxInstallAttempts: RushConstants.defaultMaxInstallAttempts,
-      pnpmFilterArguments: [],
-      checkOnly: false
-    };
-    const installManager: BaseInstallManager = InstallManagerFactory.getInstallManager(
-      this._rushConfiguration,
-      this._rushGlobalFolder,
-      purgeManager,
-      installManagerOptions
     );
 
     console.log();
-    const updatePackages: Record<string, string> = {};
-    for (const packageName of packageNames) {
-      const initialVersion: string | undefined = initialVersions.get(packageName);
-      const rangeStyle: SemVerStyle = rangeStyles.get(packageName)!;
+    const commonVersions: CommonVersionsConfiguration = this._rushConfiguration.getCommonVersions(variant);
+    const dependenciesToAddOrUpdate: Record<string, string> = {};
+    for (const { packageName, version: initialVersion, rangeStyle } of packagesToAdd) {
+      const existingSpecifiedVersions: Set<string> | undefined = allDependencyVersions.get(packageName);
+      let implicitlyPreferredVersion: string | undefined;
+      if (existingSpecifiedVersions?.size === 1) {
+        implicitlyPreferredVersion = existingSpecifiedVersions.values().next().value;
+      }
```

Comment:
`next()` is a plumbing API for iterators -- it looks a little odd to call it from business logic.

`Array.from(existingSpecifiedVersions)[0]`?  

Another idea might be to introduce a `node-core-library` helper like `Iterate.first(iterator)`, `Iterate.getRange(iterator, start, count)`, ...

Reply:
It felt silly to allocate an array to just get a singleton value out of it, but performance probably isn't super critical here. Changed to the former.